### PR TITLE
daemon/config: change DNSConfig.DNS to a []net.IP

### PR DESCRIPTION
**Pull request by thaJeztah**
- follow-up to / related to: https://github.com/moby/moby/pull/46801

Use a strong type for the DNS IP-addresses so that we can use flags.IPSliceVar, instead of implementing our own option-type and validation.

Behavior should be the same, although error-messages have slightly changed:

Before this patch:

    dockerd --dns 1.1.1.1oooo --validate
    Status: invalid argument "1.1.1.1oooo" for "--dns" flag: 1.1.1.1oooo is not an ip address
    See 'dockerd --help'., Code: 125

    cat /etc/docker/daemon.json
    {"dns": ["1.1.1.1"]}

    dockerd --dns 2.2.2.2 --validate
    unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives are specified both as a flag and in the configuration file: dns: (from flag: [2.2.2.2], from file: [1.1.1.1])

    cat /etc/docker/daemon.json
    {"dns": ["1.1.1.1oooo"]}

    dockerd --validate
    unable to configure the Docker daemon with file /etc/docker/daemon.json: merged configuration validation from file and command line flags failed: 1.1.1.1ooooo is not an ip address

With this patch:

    dockerd --dns 1.1.1.1oooo --validate
    Status: invalid argument "1.1.1.1oooo" for "--dns" flag: invalid string being converted to IP address: 1.1.1.1oooo
    See 'dockerd --help'., Code: 125

    cat /etc/docker/daemon.json
    {"dns": ["1.1.1.1"]}

    dockerd --dns 2.2.2.2 --validate
    unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives are specified both as a flag and in the configuration file: dns: (from flag: [2.2.2.2], from file: [1.1.1.1])

    cat /etc/docker/daemon.json
    {"dns": ["1.1.1.1oooo"]}

    dockerd --validate
    unable to configure the Docker daemon with file /etc/docker/daemon.json: invalid IP address: 1.1.1.1oooo

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Interesting; looks like we're not very consistent with "defaults" configured on the daemon. For log-driver (and options), we bake the defaults into the container when it's created, but looks like we're not doing that for DNS; i.e. if the daemon was started with `--dns 1.1.1.1`, then the container _does_ use `1.1.1.1` as DNS;

```
docker run --rm busybox cat /etc/resolv.conf
nameserver 1.1.1.1
```

But inspecting a container won't show that in the config;

```
docker inspect --format '{{.HostConfig.Dns}}' defaultdns
[]
```

Whereas manually specifying the `--dns` does get baked into the container;

```
docker create --dns 2.2.2.2 --name customdns busybox
docker inspect --format '{{.HostConfig.Dns}}' customdns
[2.2.2.2]
```

I _guess_ this was by design, but it's not very consistent.

File: daemon/config/config_test.go [hunk 1]
```diff
@@ -534,6 +502,34 @@ func TestValidateConfiguration(t *testing.T) {
...
+
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.doc, func(t *testing.T) {
+			var cfg Config
+			err := json.Unmarshal([]byte(tc.input), &cfg)
+			assert.Check(t, is.Error(err, tc.expectedErr))
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply daemon/config/config_test.go:531]: It tests _slighty_ more than just the stdlib as it'd fail if the `dns` field was renamed, or if it was changed from a slice to something else. But yeah, honestly it has pretty much no value.

File: daemon/container_operations.go [hunk 2]
```diff
@@ -726,7 +734,7 @@ func (daemon *Daemon) connectToNetwork(cfg *config.Config, container *container.
 
 	// TODO(thaJeztah): should this fail early if no sandbox was found?
 	sb, _ := daemon.netController.GetSandbox(container.ID)
-	createOptions, err := buildCreateEndpointOptions(container, n, endpointConfig, sb, cfg.DNS)
+	createOptions, err := buildCreateEndpointOptions(container, n, endpointConfig, sb, ipAddresses(cfg.DNS))
```
Comment (akerouanton) [inline daemon/container_operations.go:737]: And same here; it qualifies for a follow-up clean up PR.

**Review by akerouanton — APPROVED** [review]
LGTM

File: daemon/container_operations.go [hunk 3]
```diff
@@ -49,7 +57,7 @@ func (daemon *Daemon) buildSandboxOptions(cfg *config.Config, container *contain
 	if len(container.HostConfig.DNS) > 0 {
 		sboxOptions = append(sboxOptions, libnetwork.OptionDNS(container.HostConfig.DNS))
 	} else if len(cfg.DNS) > 0 {
-		sboxOptions = append(sboxOptions, libnetwork.OptionDNS(cfg.DNS))
+		sboxOptions = append(sboxOptions, libnetwork.OptionDNS(ipAddresses(cfg.DNS)))
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply daemon/container_operations.go:60]: Yes, I initially had that, but line 58 above takes a slice of strings as input 😞, so it was a case of "win some" / "lose some", so I reverted it back to accepting strings.

**Review by neersighted — APPROVED** [review]

**Comment by thaJeztah** [comment]
> Interesting; looks like we're not very consistent with "defaults" configured on the daemon. For log-driver (and options), we bake the defaults into the container when it's created, but looks like we're not doing that for DNS; i.e. if the daemon was started with `--dns 1.1.1.1`, then the container _does_ use `1.1.1.1` as DNS;


- opened a tracking ticket for the above; https://github.com/moby/moby/issues/46822

**Pull request merged**